projects/bb3/agents/opt_bb3_agent.py

```diff
@@ -285,27 +274,65 @@ def get_orm_observation(
         """
         agent = self.agents[Module.OPENING_DIALOGUE]
         agent.reset()
-        for i, mem in enumerate(opening_memories):
+        prefixed_memories = {}
+        for mem, val in opening_memories.items():
             mem = MemoryUtils.maybe_add_memory_prefix(mem, 'partner', self.MODEL_TYPE)
-            opening_memories[i] = mem
+            prefixed_memories[mem] = val
 
         new_obs = copy.deepcopy(observation)
-        new_obs.force_set(
-            'text', self._check_and_limit_len('\n'.join(opening_memories))
+        memories_to_use = MemoryUtils.get_available_memories(
+            '',
+            prefixed_memories,
+            set(),
+            dictionary=self.dictionary,
+            **self._get_memory_heuristic_values(),
         )
-        new_obs.force_set('memories', opening_memories)
+        if not memories_to_use:
+            # we need at least one memory to open with...
+            memories_to_use = random.sample(list(prefixed_memories.keys()), 1)
+
+        new_obs.force_set('text', self._check_and_limit_len('\n'.join(memories_to_use)))
+        new_obs.force_set('memories', prefixed_memories)
 
         return agent.observe(new_obs)
 
+    def get_opening_memories(
+        self, memories: Optional[Union[List[str], Dict[str, int]]]
+    ) -> Optional[Dict[str, int]]:
+        """
+        Get the opening memories, if applicable.
+
+        This function is designed to handle legacy cases where memories are
+        presented as a list, rather than a dict.
+
+        :param memories:
+            memories from the opening message
+
+        :return opening_memories:
+            return the set of true opening memories
+        """
+        opening_memories = None
+        if memories:
+            if isinstance(memories, dict):
+                opening_memories = memories
+            elif isinstance(memories, list):
+                opening_memories = {}
+                for mem in memories:
+                    opening_memories = MemoryUtils.add_memory(mem, opening_memories)
+
+        assert not opening_memories or isinstance(opening_memories, dict)
+        return opening_memories
+
     def observe(self, observation: Message) -> Dict[Module, Message]:
         # handle passed memories as well
         observation = Message(observation)
-        opening_memories = observation.pop('memories', None)
+        opening_memories = self.get_opening_memories(observation.pop('memories', None))
         observations = super().observe(observation)
         for m in Module.dialogue_modules():
             ag_obs = copy.deepcopy(observation)
             observations[m] = self.agents[m].observe(ag_obs)
         if is_opener(observation['text'], opening_memories):
+            assert opening_memories
```

Comment:
Question: having no memory does it make it generate an opening message anyway? I mean, if it doesn't crash the model, it may be a legitimate case to have here (?).